Query Cache: Assume UDFs are non-deterministic

Ergus opened the pull request:
As issue #77553 explains, all UDFs were wrongly treated as deterministic opposite to the [QueryCache documentation](https://clickhouse.com/docs/en/operations/query-cache#configuration-settings-and-usage).

This PR solves that issue and adds 2 new tests to assert that this errors never pass silently again.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The query cache now assumes that UDFs are non-deterministic. Accordingly, results of queries with UDFs are no longer cached. Previously, users were able to define non-deterministic UDFs whose result would erronously be cached (issue #77553).

Ergus committed:
Consider UDF non-deterministic.

Extend the HasNonDeterministicFunctionsMatcher to check also UDFs instead of only user defined functions.

* QueryResultCache.cpp: (HasNonDeterministicFunctionsMatcher::visit) : Try harder with UserDefinedSQLFunctionFactory and
UserDefinedExecutableFunctionFactory if the functon is not deterministic.

Fixes #77553

Todo: The two new tryGet functions need to take a lock so this change could potentially impact the command processing
when the server is used concurrently and the users are sending SQLs with many UDFs.

Ergus committed:
Add tests for UDF SQL defined are non-deterministic by default.

* tests/queries/0_stateless/02494_query_cache_udf_sql.sql:
* tests/queries/0_stateless/02494_query_cache_udf_sql.reference: New files

Ergus committed:
Add tests for UDF externally are non-deterministic by default.

* tests/queries/0_stateless/02494_query_cache_udf_executable.sh:
* tests/queries/0_stateless/02494_query_cache_udf_executable.reference: New files

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=77633) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=77633) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=77633) it.</sub>

Ergus commented:
@rschu1ze the HW change needs this approved before.

Ergus commented:
Hi @yariks5s

I see that the style check is failing. Do you have some style guideline I can consult?

Is it declaring the variable inside the `if (const auto var = smething)`,
or the `} else if (const auto var = smething) {` or both?

BTW, the total workflow seems to have been executed but freeze in the complete job step. Is this intended?

yariks5s commented:
@Ergus you can go to the Style check link in the CI, and click on the `FAIL` text. After that, you can see the fail report:
<img width="1412" alt="image" src="https://github.com/user-attachments/assets/ae25332f-d04c-4b56-9eed-a3214f6450a3" />

According to our style check you should put figure brackets at the new line. You left figure brackets on the same line on: 
- /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:61
- /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:64
- /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:66

Ergus commented:
Hi @yariks5s: 

> According to our style check you should put figure brackets at the separate line. You left figure brackets on the same line on:
> 
> * /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:61
> * /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:64
> * /ClickHouse/src/Interpreters/Cache/QueryResultCache.cpp:66

Thanks, I already fixed it ;)

 I was actually asking for a clang format file or something similar where you define the rules so I can check them and setup my editor.
This seems similar to Allman but with spaces instead of tabs... But I suppose that there may be more details.

Ergus commented:
Hi @rschu1ze:

I see that some tests are failing. I tried the stateless one locally and it is failing on master, so seem unrelated with my change.

What's your common method to proceed? In an organization I would try to git-blame and try to contact the developer of the failing code in order to ask for help and get a faster solution.

rschu1ze committed:
Minor fixups

rschu1ze committed:
Minor fixups, pt. II

rschu1ze commented:
@Ergus I added a few finishing touches to the PR, it looks good. Thanks!

> I was actually asking for a clang format file or something similar where you define the rules so I can check them and setup my editor.
> This seems similar to Allman but with spaces instead of tabs... But I suppose that there may be more details.

Yeah, there is a `.clang-format` in the root directory of the repository. Though in practice, most people run a lot more often into the typo checker than the style checker :-)

> I see that some tests are failing. I tried the stateless one locally and it is failing on master, so seem unrelated with my change.

Yes, the failures are 100% unrelate4d. ClickHosue's CI is flaky and it is an uphill battle to stabilize it. 

> What's your common method to proceed? In an organization I would try to git-blame and try to contact the developer of the failing code in order to ask for help and get a faster solution.

No need to something from your side for now (though `git-blame` would make sense) - Most of the failures are tracked as issues already and being taken care of.

rschu1ze commented on tests/queries/0_stateless/02494_query_cache_udf_executable.sh:
```diff
@@ -0,0 +1,35 @@
+# Test for issue #77553: External UDFs may be non-deterministic. The query cache should treat them as such, i.e. reject them.
+# Also see 02494_query_cache_udf_sql.sql
+
+set -e
```
As per your findings in https://github.com/ClickHouse/ClickHouse/issues/59988#issuecomment-2727041022 (thanks for noticing!), we'll need to go for an integration test actually. I tried quickly with standard `clickhouse-client` instead of `clickhouse-local` but could not get it to work easily.

Re integration tests: They can be a bit more difficult to setup, see tests/integration/README.md for prerequisites. I recommend testing with the runner script.

Ergus committed:
Integration tests to check UDF's are non-deterministic by default

* tests/integration/test_executable_user_defined_function/test.py (test_executable_function_bash_cache) : New test for
executable UDFs checking the use of the cache
* tests/integration/test_sql_user_defined_functions/test.py: New file
(test_sql_user_defined_functions_cache) : New test similar to above.

Ergus commented:
Hi:

Just pushed the integration tests. I think it is sensible to delete the previous query tests as they are not doing the job they were supposed to do.

If you give green light I will revert e437ab4b3e4 and 55bbf428196; unless you think that they have some value if modified somehow.

rschu1ze commented:
@Ergus Yes, let's revert these

Ergus committed:
Revert "Add tests for UDF SQL defined are non-deterministic by default."

This reverts commit f8103cb96a4a58e7eec8322d8cf34b26e79bfce5.

Ergus committed:
Revert "Add tests for UDF externally are non-deterministic by default."

This reverts commit 0133ef6431a7ba33773ae43679ff13568972c219.

rschu1ze committed:
Fixups, pt. III

rschu1ze commented:
@Ergus While checking, I noticed that we can at least retain the stateless tests for SQL-defined UDFs and I did just that. The reason is that integration tests are a pain in the ass overall, and we rather have only the minimally required amount of them.

rschu1ze committed:
Fix test flakiness
